File: project_manage_student/Manage_Student/App/static/js/teacher_score.js
document.addEventListener('DOMContentLoaded',function(){
    const input_score = document.querySelectorAll('input[type="number"]');
    for( ip of input_score ){
        if(ip.value != '')
            ip.disabled = true;
    }
});

function get_info_by_class(){

    const class_ = document.getElementById('class-select');
    fetch('/api/teacher_score/get_info_by_class',{
        method:'post',
        body:JSON.stringify({
            'class_id':class_.value
        }),
        headers:{
           'Content-Type' : 'application/json'
        }
    }).then(res => res.json()).then(data =>{
        console.log(data)
        const subject_id = document.getElementById('subject-id');
        const semester_name = document.getElementById('semester-name');
        const semester_year = document.getElementById('semester-year');

        subject_id.innerHTML = '';
        semester_year.innerHTML = '';

        semester_year.value = `${data.year}-2025`;

        data.subjects.forEach(dt => {
            const option  = document.createElement('option');
            option.value = `${dt.subject_id}`;
            option.textContent = `${dt.subject_name}`;
            subject_id.append(option);
        })
        subject_id.addEventListener('change',function(){
                semester_name.innerHTML = '';
                data.values[this.value].forEach( val =>{
                const option = document.createElement('option');
                option.value = `${val.semester_id}`;
                option.textContent = `${val.semester_name}`;

                semester_name.append(option);
            } )
        });
        subject_id.dispatchEvent(new Event('change'));
    });
}

function saveScore(teacher_plan_id){
    const score=[];
    const input_score = document.querySelectorAll('input[type="number"]');

    for( ip of input_score ){
        score.push({
            'student_id':ip.dataset.id,
            'type_exam':ip.className.slice(0,-2),
            'value':ip.value,
            'count_exam':ip.className.slice(-1)
        });
    }

    fetch(`/api/teacher_score/${teacher_plan_id}/score`,{
        method:'post',
        body:JSON.stringify({
            'list_score':score
        }),
        headers:{
           'Content-Type' : 'application/json'
        }
    }).then(res => res.json()).then(data =>{
        if (data.status === 200) {
             confirm("Lưu thành công!");
             window.location.href = `/teacher/view_score/${teacher_plan_id}`;
        }
    });
}

function editScore(student_id,teacher_plan_id){
    const input_score = document.querySelectorAll(`input[data-id="${student_id}"]`);
    for( ip of input_score ){
        ip.disabled = ! ip.disabled;
    }

    const div_action = document.querySelector(`div.action[data-id="${student_id}"]`);
    div_action.innerHTML = `
                <button class="btn btn-primary" onclick="confirm_edit(${student_id},${teacher_plan_id})">Xác nhận</button>
                <button class="btn btn-danger" onclick="cancel_edit(${student_id},${teacher_plan_id})">Hủy</button>
    `;
}

function confirm_edit(student_id,teacher_plan_id){
    const input_score = document.querySelectorAll(`input[data-id="${student_id}"]`);
    score=[];
    for( ip of input_score ){
        if(ip.value != ''){
            score.push({
                'student_id':ip.dataset.id,
                'type_exam':ip.className.slice(0,-2),
                'value':ip.value,
                'count_exam':ip.className.slice(-1)
             });
        }
    }
    fetch(`/api/teacher_score/${teacher_plan_id}/edit_score`,{
        method:'put',
        body:JSON.stringify({
            'list_score':score
        }),
        headers:{
           'Content-Type' : 'application/json'
        }
    }).then(res => res.json()).then(data =>{
        console.log(data)
        if (data.status === 200) {
             confirm("Đã sửa điểm thành công!");
             window.location.href = `/teacher/teacher_input_score/${teacher_plan_id}`;
        }
    });
}

function cancel_edit(student_id,teacher_plan_id){
     const input_score = document.querySelectorAll(`input[data-id="${student_id}"]`);
        for( ip of input_score ){
            ip.disabled = ! ip.disabled;
        }

    const div_action = document.querySelector(`div.action[data-id="${student_id}"]`);
    div_action.innerHTML = `
               <button class="btn btn-danger" onclick="editScore(${student_id},${teacher_plan_id})">Sửa</button>
    `;
    window.location.reload();
}

File: project_manage_student/Manage_Student/App/api/teacher_score.py
import pdb
from datetime import datetime

from flask_login import current_user

from App import app, db
from App.dao import teacher_list,exam
from flask import jsonify, request
from App.model import Exam, Score


@app.route('/api/teacher_score/get_info_by_class', methods=['POST'])
def get_info_by_class():
    class_id = request.json.get('class_id')
    teach_class = teacher_list.get_plan_by_class_id(class_id=class_id, teacher_id=current_user.id)

    teach_class_json = {}
    subjects = []
    for t in teach_class:
        if t.teacher_subjects.subjects.id not in teach_class_json:
            teach_class_json[t.teacher_subjects.subjects.id] = []
            subjects.append({
                'subject_id': t.teacher_subjects.subjects.id,
                'subject_name': t.teacher_subjects.subjects.name,
            })

        teach_class_json[t.teacher_subjects.subjects.id].append({
            'subject_id': t.teacher_subjects.subjects.id,
            'semester_id': t.semester_id,
            'semester_name': t.semester.name.name
        })
    return jsonify({'values': teach_class_json, 'year': datetime.now().year,'subjects':subjects})


@app.route('/api/teacher_score/<int:teacher_plan_id>/score', methods=['GET','POST'])
def add_score(teacher_plan_id):
    list_score = request.json.get('list_score')

    for s in list_score:
        if s.get('student_id') and teacher_plan_id:
            tmp_exam = exam.get_exam_by_student_id(teacher_plan_id=teacher_plan_id,student_id=int(s['student_id']))
            if tmp_exam is None:
                tmp_exam = Exam(teacher_plan_id=teacher_plan_id, student_id=int(s['student_id']))
                db.session.add(tmp_exam)
                db.session.commit()
            if s.get('type_exam') and s.get('value'):
                is_score_exit = (Score.query.filter(Score.type_exam.__eq__(s.get('type_exam')),
                                                    Score.exam_id.__eq__(tmp_exam.id),
                                           Score.count_exam.__eq__(s.get('count_exam'))).first())
                if is_score_exit is None:
                    new_score = Score(type_exam=s.get('type_exam'),
                                      value=float(s.get('value')),
                                      count_exam=s.get('count_exam'),
                                      exam_id=tmp_exam.id)
                    db.session.add(new_score)
    db.session.commit()
    return jsonify({'status': 200})


@app.route('/api/teacher_score/<int:teacher_plan_id>/edit_score', methods=['PUT'])
def edit_score(teacher_plan_id):
    list_score = request.json.get('list_score')
    for s in list_score:
        if s.get('student_id') and teacher_plan_id:
            tmp_exam = exam.get_exam_by_student_id(teacher_plan_id=teacher_plan_id,student_id=int(s['student_id']))
            if s.get('type_exam') and s.get('value'):
                score = (Score.query.filter(Score.exam_id.__eq__(tmp_exam.id),
                                           Score.type_exam.__eq__(s.get('type_exam')),
                                           Score.count_exam.__eq__(s.get('count_exam')))
                         .first())
                score.value = float(s.get('value'))
    db.session.commit()
    return jsonify({'status': 200})